# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [12]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest
import scipy.stats as st
from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [7]:
# Run this code:
import pandas as pd

df = pd.read_csv('../data/Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [98]:
# Your code here
df.head(20)

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86
5,"ABBASI, CHRISTOPHER",STAFF ASST TO THE ALDERMAN,CITY COUNCIL,F,Salary,NaN,50436.0,NaN
6,"ABBATACOLA, ROBERT J",ELECTRICAL MECHANIC,AVIATION,F,Hourly,40.0,NaN,46.10
7,"ABBATE, JOSEPH L",POOL MOTOR TRUCK DRIVER,STREETS & SAN,F,Hourly,40.0,NaN,35.60
8,"ABBATEMARCO, JAMES J",FIRE ENGINEER-EMT,FIRE,F,Salary,NaN,103350.0,NaN
9,"ABBATE, TERRY M",POLICE OFFICER,POLICE,F,Salary,NaN,93354.0,NaN


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [ ]:
# Your code here
sample_df = df[df["Salary or Hourly"]=="Hourly"]['Hourly Rate']
control_value = 30
alpha = 0.05 #95% confidence interval
p_value = st.ttest_1samp(sample_df,
                         control_value).pvalue #p value of the two-sided test

h_zero = f"the mean of the sample is equal to {control_value}"
h_one = f"the mean of the sample is not equal to {control_value}"
h_zero_rejected = p_value < alpha
print(p_value)
print(f"H0 rejected: {h_zero_rejected}")
if h_zero_rejected:
    print(h_one)
else:
    print(h_zero)


4.323024048623005e-92
H0 rejected: True
the mean of the sample is not equal to 30


# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [96]:
# Your code here
from scipy.stats import t
degrees_freedom = len(sample_df)-1 #degrees of freedom are equal to number of rows -1
sample_mean = sample_df.mean()
standard_error = sem(sample_df, axis=0, ddof=1, nan_policy='propagate', keepdims=False)

conf_int = list(map(float,t.interval(
    confidence = 0.95, #confidence interval
    df = degrees_freedom,
    loc = sample_mean,
    scale = standard_error
    )))

print(f"95% confidence interval: {conf_int}")

95% confidence interval: [32.52345834488425, 33.05365708767623]


### interpretation
We are 95% confident that the true hourly wages mean for the hourly workers lies between $32.52 and $33.05. 

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).



## Hypothesis Testing for Proportions

Hypothesis testing for proportions involves determining whether the proportion of success in a sample is significantly different from a hypothesized proportion. This is commonly used in scenarios where you're interested in categorical outcomes, such as polling, quality control, etc.

### Key Concepts

- **Proportion**: A proportion represents a part of a whole and is often denoted by \( p \).

- **Null Hypothesis (\(H_0\))**: Assumes that the sample proportion is equal to the hypothesized population proportion (\( p_0 \)).

- **Alternative Hypothesis (\(H_1\))**: Suggests that the sample proportion differs from the hypothesized population proportion.

### Practical Example using SciPy

Let's say we want to test if the proportion of defective items in a batch is different from 5% (0.05):

1. **Hypotheses**:
   - \(H_0\): \( p = 0.05 \)
   - \(H_1\): \( p \neq 0.05 \)

2. **Data**: Suppose we have a sample of 200 items with 15 defective ones.

Here's how to conduct the hypothesis test using `scipy.stats`:



In [97]:
from statsmodels.stats.proportion import proportions_ztest

# Data
n = 200  # Total number of items
successes = 15  # Number of defective items
p_null = 0.05  # Hypothesized population proportion

# Conducting the z-test
stat, p_value = proportions_ztest(count=successes, nobs=n, value=p_null)

# Results
print(f"Statistic: {stat}, P-value: {p_value}")

# Interpretation
alpha = 0.05  # Significance level
if p_value < alpha:
    print("Reject the null hypothesis. The proportion is significantly different from 0.05.")
else:
    print("Fail to reject the null hypothesis. The proportion is not significantly different from 0.05.")

Statistic: 1.3423121104280482, P-value: 0.17949481837303083
Fail to reject the null hypothesis. The proportion is not significantly different from 0.05.


In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [ ]:
# Your code here

# Data
hourly_df = df[df["Salary or Hourly"]=="Hourly"] #sample hourly workers
n = len(df)  # Total workers

h_workers = len(hourly_df)  # Number of hourly workers
p_null = 0.25  # Hypothesis we want to refuse

# Conducting the z-test
stat, p_value = proportions_ztest(count=h_workers, nobs=n, value=p_null)

# Results
print(f"Statistic: {stat}, P-value: {p_value}")

# Interpretation
alpha = 0.05  # Significance level
print(f"the hourly workers proportion in the sample is: {round(h_workers/n,2)}")
if p_value < alpha:
    print(f"Reject the null hypothesis. The proportion is signicantly different from {p_null}")
else:
    print(f"Fail to reject the null hypothesis. The proportion is not different from {p_null}")

Statistic: -3.5099964213703005, P-value: 0.0004481127249057967
the hourly workers proportion in the sample is: 0.24
Reject the null hypothesis. The proportion is different from 0.25
